src/python/WMCore/JobStateMachine/ChangeState.py

```diff
@@ -302,7 +305,23 @@ def recordInCouch(self, jobs, newstate, oldstate, updatesummary = False):
                 logging.debug("Updated job summary state history for job %s" % jobSummaryId)
 
             if job.get("fwjr", None):
-
+                cached_task = self.task_cache.get(job['task'])
+                if cached_task is None:
+                    spec_path = self.getWorkflowSpecDAO.execute(job['task'])['spec']
```

Comment:
ah yes just read your post above with the suggested cache structures in. I was going for something like:

```python
{'task1': {'campaign': blah, 'prepid': blah}, 'task2': {'campaign': blah, 'prepid': blah}}
```

but actually I think I like your first suggestion maybe better as there is less repetition. The only think is that the workflow information is unnecessary which is why in my example I was flattening the workflows and ignoring them.